File: game.py
import random


def roll_dice() -> int:
    return random.randint(1, 6)


class Piece:
    def __init__(self, index):
        self.index = index
        self.position = None


class Player:
    def __init__(self, color, start_position):
        self.color = color
        self.start_postion = start_position
        self.pieces = [Piece(i) for i in range(4)]

    def move_piece(self, steps):
        for piece in self.pieces:
            # If a piece is on the starting position, it must be moved away first
            if self.start_postion in [piece.position for piece in self.pieces]:
                if piece.position == self.start_postion:
                    print(
                        f"Player {self.color} must move piece {piece.index} away from start position"
                    )
                    piece.position += steps
                    return

            # If rolled a six, piece must be moved out of the starting area
            if steps == 6 and None in [piece.position for piece in self.pieces]:
                if piece.position is None:
                    print(
                        f"Player {self.color} rolls a six and moves piece {piece.index} out"
                    )
                    piece.position = self.start_postion
                    return

        # Find most advanced piece
        best_piece = None
        for piece in self.pieces:
            if best_piece is None:
                best_piece = piece
            if piece.position is not None and piece.position > best_piece.position:
                best_piece = piece

        # Move most advanced piece
        if best_piece.position is not None:
            print(f"Player {self.color} moves {steps}")
            best_piece.position += steps
            return

        # Can't move any piece
        print(f"Player {self.color} can't move a piece")


class Board:
    def __init__(self):
        # The total length of the board is one line of 8 spots per player (8*6 = 48)
        self.length = 48
        self.colors = ["black", "blue", "red", "green", "purple", "yellow"]
        self.fields = [None] * self.length
        self.players = [Player(color, i * 8) for i, color in enumerate(self.colors)]

    def game_state(self):
        return f"Players: {', '.join([player.color for player in self.players])}"

    def play(self):
        roll_count = 0
        rounds = 1
        game_over = False
        while not game_over:
            for player in self.players:
                print(f"{player.color} is playing")
                dice = roll_dice()
                roll_count += 1
                print(f"Player {player.color} rolls a {dice}")
                player.move_piece(dice)
                rounds += 1

                for piece in player.pieces:
                    if (
                        piece.position is not None
                        and (piece.position - player.start_postion) > self.length
                    ):
                        print(f"Player {player.color} wins")
                        game_over = True
                        break

                if game_over:
                    print(f"Game over after {roll_count} dice rolls")
                    print(f"Game over after {rounds} rounds")
                    break

File: six-player-ludo-csharp/SixPlayersLudo/Game.cs
namespace SixPlayersLudo;

public class Game
{
    private readonly Random _rand;

    // The total length of the board is one line of 8 spots per player (8*6 = 48)
    private const int BoardSize = 48;
    private static readonly string[] Colors = ["black", "blue", "red", "green", "purple", "yellow"];
    private readonly List<Player> _players = [];
    private readonly Piece?[] _mainPath = new Piece?[BoardSize];

    public Game()
    {
        _rand = new Random();
        var startPosition = 1;
        foreach (var color in Colors)
        {
            _players.Add(new Player(color, startPosition));
            startPosition += 8;
        }
    }

    public int RollDice()
    {
        return _rand.Next(1, 7);
    }

    public void Play()
    {
        var gameEnded = false;
        var rounds = 0;
        while (!gameEnded)
        {
            ShowGameState();

            foreach (var player in _players)
            {
                gameEnded = MakeAMove(player);
                if (gameEnded)
                {
                    Console.WriteLine("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!");
                    Console.WriteLine($"Player {player.PlayerColor} has won after {rounds} Rounds!");
                    Console.WriteLine("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!");
                    return;
                }
            }

            rounds++;
        }
    }

    private bool MakeAMove(Player player)
    {
        var result = RollDice();
        Console.WriteLine($"{player.PlayerColor} rolled {result}");

        // Move Piece
        if (result == 6 && player.CanMovePieceOutOfHome())
        {
            var piece = player.GetPieceToMoveOut();
            _mainPath[player.StartPosition] = piece;
            if (piece != null)
            {
                piece.Position = player.StartPosition;
            }
        }
        else if (player.CanMoveForwardInGoal(result))
        {
            player.MoveForwardInGoal(result);
            Console.WriteLine($" {player.PlayerColor} moved forward inside goal");
        }
        else if (player.HasPieceOnBoard())
        {
            var pieceIntoGoal = player.GetPieceToMoveIntoGoal(result);
            if (pieceIntoGoal != null)
            {
                _mainPath[pieceIntoGoal.Position] = null;
            }
            else
            {
                var piece = player.GetMostAdvancedPiece();
                if (piece != null)
                {
                    _mainPath[piece.Position] = null;
                    var newPosition = (piece.Position + result) % BoardSize;

                    var thrownPiece = _mainPath[newPosition];
                    if (thrownPiece != null)
                    {
                        piece.Position = 0;
                    }

                    _mainPath[newPosition] = piece;
                    piece.Position = newPosition;
                }
            }
        }

        if (player.HasWon())
        {
            return true;
        }

        // Roll Dice Again
        if (result == 6)
        {
            Console.WriteLine($"{player.PlayerColor} is allowed to roll the dice again");
            MakeAMove(player);
        }

        return false;
    }

    private void ShowGameState()
    {
        Console.WriteLine("------------------------------------------------------------------------------------");
        for (var i = 0; i < BoardSize; i++)
        {
            var piece = _mainPath[i];
            if (piece != null)
            {
                Console.WriteLine($"Piece {piece.PieceName} is in position {i}");
            }
        }
    }
}